client/proxy-dbus.cc

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright (c) 2016 SUSE LLC
+ *
+ * All Rights Reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of version 2 of the GNU General Public License as published
+ * by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, contact Novell, Inc.
+ *
+ * To contact Novell about this file by physical or electronic mail, you may
+ * find current contact information at www.novell.com.
+ */
+
+
+#include <sstream>
+
+#include "proxy-dbus.h"
+#include "commands.h"
+#include "utils/text.h"
+
+
+using namespace std;
+
+
+ProxySnapshotDbus::ProxySnapshotDbus(ProxySnapshotsDbus* backref, unsigned int num)
+    : num(num)
+{
+    XSnapshot x = command_get_xsnapshot(*backref->backref->backref->conn,
+					backref->backref->config_name, num);
+
+    type = x.type;
+    date = x.date;
+    uid = x.uid;
+    pre_num = x.pre_num;
+    description = x.description;
+    cleanup = x.cleanup;
+    userdata = x.userdata;
+}
+
+
+ProxySnapshotDbus::ProxySnapshotDbus(SnapshotType type, unsigned int num, time_t date, uid_t uid,
+				     unsigned int pre_num, const string& description,
+				     const string& cleanup, const map<string, string>& userdata)
+    : type(type), num(num), date(date), uid(uid), pre_num(pre_num), description(description),
+      cleanup(cleanup), userdata(userdata)
+{
+}
+
+
+void
+ProxySnapperDbus::setConfigInfo(const map<string, string>& raw)
+{
+    command_set_xconfig(*backref->conn, config_name, raw);
+}
+
+
+ProxySnapshots::const_iterator
+ProxySnapperDbus::createSingleSnapshot(const SCD& scd)
+{
+    unsigned int num = command_create_single_xsnapshot(*backref->conn, config_name, scd.description,
+						       scd.cleanup, scd.userdata);
+
+    proxy_snapshots.emplace_back(new ProxySnapshotDbus(&proxy_snapshots, num));
+
+    return --proxy_snapshots.end();
```

Comment:
When I first saw this, it looked somehow wrong to me: Take the iterator that points one element beyond the last element of the container and go one element back.

But it looks like STL containers really can't do that any better: vector::back() returns a reference to the last element, not an iterator; vector::rbegin() returns a reverse iterator, not a forward iterator as everybody using such a function would expect. So this is indeed the best solution STL has to offer.